pkgs/tools/filesystems/bees/bees-service-wrapper

```diff
@@ -0,0 +1,104 @@
+
+[[ $bees_debug ]] && { PS4=':${BASH_SOURCE##*/}:$LINENO+'; set -x; }
+
+usage() {
+  cat >&2 <<EOF
+Usage: ${BASH_SOURCE##*/} run|cleanup config-name|fs-spec[;idxSizeMB=...][;verbosity=...][;workDir=...]
+
+  fs-spec should be in a format recognized by find-mnt.
+EOF
+  exit 1
+}
+
+die() { echo "$*" >&2; exit 1; }
+
+do_setup() {
+  declare -n options_src=$1; shift
+  declare fs_spec=$1; shift
+  declare -g uuid bees_{work,run,mnt}_dir bees_verbosity bees_db_size{,_mb} blockdev
+  for key in "${!options_src[@]}"; do
+    case $key in
+      workDir) bees_work_dir=${options_src[$key]} ;;
+      verbosity) bees_verbosity=${options_src[$key]}  ;;
+      idxSizeMB) bees_db_size_mb=${options_src[$key]} ;;
+    esac
+  done
+
+  : bees_config_dir="${bees_config_dir:=/etc/bees}"
+  if [[ -s $bees_config_dir/$fs_spec.conf ]]; then
+    source "$bees_config_dir/$fs_spec.conf"
+  fi
+  read -r uuid fstype < <(findmnt -n -o uuid,fstype "$fs_spec") && [[ $fstype ]] || exit
+  for f in "${bees_config_dir}/$fs_spec.conf" "$bees_config_dir/UUID=$uuid.conf"; do
+    [[ -s $f ]] && source "$f"
+  done
+  : bees_work_dir="${bees_work_dir:=.beeshome}"
+  : bees_run_dir="${bees_run_dir:=/run/bees}"
+  : bees_mnt_dir="${bees_mnt_dir:=$bees_run_dir/mnt/$uuid}"
+  : bees_home="${bees_home:=$bees_mnt_dir/$bees_work_dir}"
+  : bees_verbosity="${bees_verbosity:-6}"
+  : bees_db_size_mb="${bees_db_size_mb:=1024}"
+  : blockdev=${blockdev:="/dev/disk/by-uuid/$uuid"}
+  bees_db_size="$(( bees_db_size_mb * 1024 * 1024 ))"
+
+  [[ -b $blockdev ]] || die "Block device $blockdev missing"
+  [[ $fstype = btrfs ]] || die "Device type is $fstype, not btrfs"
+  (( bees_db_size_mb % 16 == 0 )) || die "DB size must be divisible by 16MB"
+}
+
+do_run() {
+  local db old_db_size
+
+  do_setup "$@"
+  mkdir -p -- "$bees_mnt_dir" || exit
+  mountpoint -q "$bees_mnt_dir" || mount -osubvolid=5 -- "$blockdev" "$bees_mnt_dir" || exit
+  if [[ -d $bees_home ]]; then
+    btrfs subvolume show "$bees_home" >/dev/null 2>&1 || die "$bees_home exists but is not a subvolume"
+  else
+    btrfs subvolume create "$bees_home" || exit
+  fi
+  db=$bees_home/beeshash.dat
+  touch -- "$db"
+
+  old_db_size=$(stat -c %s -- "$db")
+  new_db_size=$bees_db_size
+
+  if (( old_db_size != new_db_size )); then
+    rm -f -- "$bees_home"/beescrawl."$uuid".dat
+    truncate -s "$new_db_size" -- "$db" || exit
+  fi
+  chmod 700 -- "$db_path"
+  exec "${beesd_bin:-/lib/bees/bees}" "$bees_mnt_dir" || exit
+}
+
+do_cleanup() {
+  do_setup "$@"
+  mountpoint -q "$bees_mnt_dir" && umount -l -- "$bees_mnt_dir" || exit
+}
+
+[[ $# = 2 ]] || usage
+declare -f "do_$1" >/dev/null 2>&1 || usage
+mode=$1; shift
+
+declare -a args=( "$1" ); shift
+declare -A options=( )
+IFS=';' read -r -a pieces <<<"$2"
+for piece in "${pieces[@]}"; do
+  if [[ $piece = *=* ]]; then
+    options[${piece%%=*}]=${piece#*=}
+  else
+    args+=( "$piece" )
+  fi
+done
+
+"do_$mode" options "${args[@]}"
```

Comment:
I don't really feel qualified to review this wrapper. Do you have maybe a link to the upstream wrapper? How much does it differ?

Reply:
It's pretty much a completely independent reimplementation; upstream is at https://github.com/Zygo/bees/blob/v0.6.1/scripts/beesd.in

Reply:
@charles-dyfis-net Ah, well in that case it's even easier: `types.addCheck types.int (n: mod n 16 == 0)`

Reply:
Excellent. I think I've addressed everything, now, except for the matter of finding someone more comfortable reviewing the rewritten wrapper.